core/src/commonMain/kotlin/org/aquila3d/core/math/Quaternion.kt

```diff
@@ -0,0 +1,324 @@
+        return this
+    }
+
+    fun divide(scale: Float): Quaternion {
+        return divide(scale.toDouble())
+    }
+
+    fun divide(scale: Double): Quaternion {
+        w /= scale
+        x /= scale
+        y /= scale
+        z /= scale
+        return this
+    }
+
+    fun normalize(): Quaternion {
+        val magnitude = magnitude()
+        return divide(magnitude)
+    }
+
+    fun invert(): Quaternion {
+        x = -x
+        y = -y
+        z = -z
+        return this
+    }
+
+    /**
+     * Rotates [vector3] by the rotation represented in this [Quaternion]. Mathematically this is equivalent to:
+     *
+     * **v'** = **q** * **v** * ~**q** where ~**q** is the conjugate or inverse of this [Quaternion].
+     */
+    fun rotate(vector3: Vector3): Vector3 {
+        val rotation = Quaternion(this)
+        val conj = Quaternion(this).invert()
+        return (rotation.multiply(vector3).multiply(conj).getAxis())
+    }
+
+    override fun toString(): String {
+        return "Quaternion(w=$w, x=$x, y=$y, z=$z)"
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other == null || this::class != other::class) return false
+
+        other as Quaternion
+
+        if (w != other.w) return false
+        if (x != other.x) return false
+        if (y != other.y) return false
+        if (z != other.z) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = w.hashCode()
+        result = 31 * result + x.hashCode()
+        result = 31 * result + y.hashCode()
+        result = 31 * result + z.hashCode()
+        return result
+    }
+
+    companion object {
+
+        /**
+         * Constructs a new [Quaternion] by applying the provided Euler rotation angles in order, yaw - pitch - roll.
+         * Note that these Euler angles are defined in GL coordinate space, not the typical rigid body mechanics space.
+         *
+         * @param yaw Rotation about the +Y axis, in [Degrees]
+         * @param pitch Rotation about the +X Axis, in [Degrees]
+         * @param roll Rotation about the +Z Axis, in [Degrees]
+         */
+        fun fromEulerAnglesDeg(yaw: Degrees, pitch: Degrees, roll: Degrees): Quaternion {
+            return Quaternion(yaw.toRadians(), pitch.toRadians(), roll.toRadians())
+        }
+
+        /**
+         * Constructs a new [Quaternion] by applying the provided Euler rotation angles in order, yaw - pitch - roll.
+         * Note that these Euler angles are defined in GL coordinate space, not the typical rigid body mechanics space.
+         *
+         * @param yaw Rotation about the +Y axis, in [Radians]
+         * @param pitch Rotation about the +X Axis, in [Radians]
+         * @param roll Rotation about the +Z Axis, in [Radians]
+         */
+        fun fromEulerAnglesRad(yaw: Radians, pitch: Radians, roll: Radians): Quaternion {
+            return Quaternion(yaw, pitch, roll)
```

Comment:
This is a style preference but generally when a method return is a single line, you inline it to reduce the method to one or two lines instead of three.

```
fun doSomething(): Other = Other(...)
```